test/e2e/apps/job.go

```diff
@@ -344,6 +344,48 @@ var _ = SIGDescribe("Job", func() {
 		}
 	})
 
+	ginkgo.It("should recreate pods only after they have failed if pod replacement policy is set to Failed", func(ctx context.Context) {
+		ginkgo.By("Creating a job")
+		job := e2ejob.NewTestJob("notTerminate", "pod-recreate-failed", v1.RestartPolicyNever, 1, 1, nil, backoffLimit)
+		failedPodReplacementPolicy := batchv1.Failed
+		job.Spec.PodReplacementPolicy = &failedPodReplacementPolicy
+		job.Spec.Template.Spec.Containers[0].Image = "gcr.io/k8s-staging-perf-tests/sleep"
+		job.Spec.Template.Spec.Containers[0].Args = []string{"-termination-grace-period", "10s", "20s"}
+		job, err := e2ejob.CreateJob(ctx, f.ClientSet, f.Namespace.Name, job)
+		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Deleting job pods")
+		err = e2ejob.WaitForJobPodsRunning(ctx, f.ClientSet, f.Namespace.Name, job.Name, parallelism)
+		framework.ExpectNoError(err, "failed to ensure number of pods associated with job %s is equal to parallelism count in namespace: %s", job.Name, f.Namespace.Name)
+
+		pods, err := e2ejob.GetJobPods(ctx, f.ClientSet, f.Namespace.Name, job.Name)
+		framework.ExpectNoError(err, "failed to get pod list for job in namespace: %s", f.Namespace.Name)
+
+		framework.ExpectNoError(e2epod.DeletePodsWithGracePeriod(ctx, f.ClientSet, pods.Items, 0), "failed to delete pods in namespace: %s", f.Namespace.Name)
```

Comment:
Actually let's redesign this test:
1. Set `-termination-grace-period` to something long (1min should be fine).
2. Change `DeletePodsWithGracePeriod` to have a grace period (not zero).
3. At this point (line 364, ), check that the Job status has one terminating pod and no active pods.
4. Once we have verified the desired status, issue another `DeletePodsWithGracePeriod`, but this time with zero grace period. This will cause a SIGKILL and immediately delete the pods.
5. Verify pods recreated.